src/cpp/database/database.cpp

```diff
@@ -2605,6 +2945,1117 @@ DatabaseDump Database::dump_data_(
+    DatabaseDump reference_container = dump.at(reference_tag);
+
+    // Check all references
+    for (auto refIt = references_id.begin(); refIt != references_id.end(); ++refIt)
+    {
+        check_entity_reference(reference_container, std::to_string(boost::lexical_cast<int>(
+                    (std::string)*refIt)), entity_tag, entity_id);
+    }
+}
+
+void Database::check_entity_reference_contains(
+        DatabaseDump const& container,
+        std::string const& reference_id,
+        std::string const& entity_tag,
+        std::string const& entity_id)
+{
+    check_entity_exists(container, reference_id);
+
+    std::vector <std::string> reference_entities = container.at(reference_id).at(entity_tag);
+
+    bool found = false;
+    size_t i = 0;
+    while (i < reference_entities.size() && !found)
+    {
+        found = (reference_entities[i] == entity_id);
+        i++;
+    }
+
+    if (!found)
+    {
+        throw CorruptedFile("Entity with ID (" + reference_id + ") :" + container.at(reference_id).dump() +
+                      " have reference to " + entity_tag + ": " +
+                      container.at(reference_id).at(entity_tag).dump() +
+                      " instead of " + entity_tag + ": " + entity_id);
+    }
+}
+
+void Database::check_contains_reference(
+        nlohmann::detail::iter_impl<nlohmann::json> const& it,
+        std::string const& entity_tag,
+        std::string const& reference_container_tag,
+        std::string const& reference_tag,
+        DatabaseDump const& dump)
+{
+    std::string entity_id = std::to_string(boost::lexical_cast<int>(it.key()));
+    std::string reference_id = std::to_string(boost::lexical_cast<int>((std::string)(*it).at(reference_tag)));
+    DatabaseDump reference_container = dump.at(reference_container_tag);
+
+    check_entity_reference_contains(reference_container, reference_id, entity_tag, entity_id);
+}
+
+void Database::check_contains_all_references(
+        nlohmann::detail::iter_impl<nlohmann::json> const& it,
+        std::string const& entity_tag,
+        std::string const& reference_tag,
+        DatabaseDump const& dump)
+{
+    std::string entity_id = std::to_string(boost::lexical_cast<int>(it.key()));
+    DatabaseDump references_id = (*it).at(reference_tag);
+    DatabaseDump reference_container = dump.at(reference_tag);
+
+    for (auto refIt = references_id.begin(); refIt != references_id.end(); ++refIt)
+    {
+        check_entity_reference_contains(reference_container,
+                std::to_string(boost::lexical_cast<int>((std::string)*refIt)), entity_tag, entity_id);
+    }
+}
+
+void Database::load_database(
+        DatabaseDump dump)
+{
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    // Hosts
+    {
+        DatabaseDump container = dump.at(HOST_CONTAINER_TAG);
+
+        // For each entity of this kind in database
+        for (auto it = container.begin(); it != container.end(); ++it)
```

Comment:
I use iterators because that is what i see in the [documentation](https://json.nlohmann.me/features/iterators/).
I agree that using a range for loop is more understandable but I don't know if it better to stay with iterators because the library exemplifies in that way.
If you think is better using a range for loop, confirm me and I will create the task.

Reply:
Task created: https://eprosima.easyredmine.com/issues/11749